File: Java/src/main/java/org/codecop/lcdnumbers/Zipper.java
package org.codecop.lcdnumbers;

import java.util.Collection;
import java.util.List;

/**
 * Zip joins elements of collections element wise, i.e. all first elements are joined and so on.
 */
public interface Zipper {

    @FunctionalInterface
    public interface Combiner<T, R> {
        R combine(T t);
    }

    <T, R> List<R> zip(Collection<? extends Collection<T>> collections, Combiner<List<T>, R> combine);

}

File: Kotlin-Spring/src/main/kotlin/org/codecop/lcdnumbers/Zipper.kt
package org.codecop.lcdnumbers

/**
 * Zip joins elements of collections element wise, i.e. all first elements are joined and so on.
 */
interface Zipper {
    @FunctionalInterface
    interface Combiner<T, R> {
        fun combine(t: T): R
    }
    
    fun <T, R> zip(collections: Collection<Collection<T>>, combine: (List<T>)-> R): List<R>
}